ipa-core/src/secret_sharing/mod.rs

```diff
@@ -57,13 +109,140 @@ pub trait Block: Sized + Copy + Debug {
-    Clone + Copy + Eq + Debug + Send + Sync + Sized + Additive + Serializable + 'static
+    Clone
+    + Copy
+    + Eq
+    + Debug
+    + Send
+    + Sync
+    + Sized
+    + Additive
+    + Serializable
+    + Vectorizable<1>
+    + 'static
 {
     type Storage: Block;
 
     const BITS: u32;
 
     const ZERO: Self;
+
+    // Note the trait bound of `Vectorizable<1>` here, i.e., these
+    // helpers only apply to arrays of a single element.
+    fn into_array<A>(self) -> A
+    where
+        Self: Vectorizable<1, Array = A>,
+        A: SharedValueArray<Self>,
+    {
+        once(self).collect::<A>()
+    }
+
+    fn from_array<A>(array: &A) -> Self
+    where
+        Self: Vectorizable<1, Array = A>,
+        A: SharedValueArray<Self>,
+    {
+        array.get(0)
+    }
+
+    fn from_array_mut<A>(array: &mut A) -> &mut Self
+    where
+        Self: Vectorizable<1, Array = A>,
+        A: SharedValueArray<Self>,
+    {
+        array.get_mut(0)
+    }
+}
+
+// Note that we can either make `trait Vectorizable<N>: SharedValue`, or we can make `trait
+// SharedValue: Vectorizable<1>`, but doing both creates a cycle. (Similarly for
+// `FieldVectorizable` / `Field`.)
+//
+// Although it is somewhat unnatural, we choose to do the latter, because it allows existing
+// high-level protocols unaware of vectorization to call vectorized versions of core protocols (with
+// width of 1) without updating all of the trait bounds. This does mean that the trait definitions
+// do not prevent implementing `Vectorizable` for something that is not a `SharedValue`, but please
+// don't do that.
+
+/// Trait for `SharedValue`s supporting operations on `N`-wide vectors.
+pub trait Vectorizable<const N: usize>: Sized {
+    type Array: SharedValueArray<Self>;
+}
+
+/// Trait for `Field`s supporting operations on `N`-wide vectors.
+///
+/// We would like `F` to be `FieldVectorizable` if it satisfies all of the following:
+///  1. `F: Field`.
+///  2. `<F as Vectorizable<N>>::Array: FieldArray<Self>`. Rust does not support expressing a
+///     constraint on a super-trait's associated type directly. Instead, this effect is achieved
+///     by constraining the `ArrayAlias` associated type and then constraining that
+///     `Vectorizable::Array == FieldVectorizable::ArrayAlias` where necessary (e.g. in the
+///     definition and blanket impl of the `FieldSimd` trait. We call it `ArrayAlias` instead of
+///     `Array` so that references to the `Array` associated type do not require qualification
+///     with a trait name.
+///  3. `F: Vectorizable<N>`. This is implied by the previous two, because `FieldArray`
+///     is a sub-trait of `SharedValueArray`.
+pub trait FieldVectorizable<const N: usize>: SharedValue + Sized {
+    type ArrayAlias: FieldArray<Self>;
+}
+
+// We could define a `SharedValueSimd` trait that is the analog of this for `SharedValue`s, but
+// there are not currently any protocols that need it.
+pub trait FieldSimd<const N: usize>:
```

Comment:
I am not sure I understand the purpose of this trait, seems to exist for the purpose of grouping various trait bounds. Is it correct?

Reply:
Exactly. It is like a convenience trait alias. Meaning something like `trait MyTrait: LongListOfTraits {}` / `impl<T: LongListOfTraits> MyTrait for T`, which lets you write `T: MyTrait` instead of `T: LongListOfTraits`.

But in this case `LongListOfTraits` is not so much a long list of traits, as a short list of traits with a complicated associated type bound. But the point is the same -- without `FieldSimd`, lots of protocols would need to say `F: Field + Vectorizable<N, Array = <Self as FieldVectorizable<N>>::ArrayAlias> + FieldVectorizable<N>`.